src/controller.rs

```diff
@@ -217,21 +232,237 @@ async fn reconcile(app_instance: Arc<AppInstance>, ctx: Arc<Context>) -> Result<
                     Action::requeue(Duration::from_secs(60))
                 }
             };
-            delete_job(&app_instance, &ctx).await?;
+            delete_job(app_instance, ctx).await?;
             action
         }
     };
 
     Ok(action)
 }
 
+async fn reconcile_delete(app_instance: &AppInstance, ctx: &Context) -> Result<Action> {
+    info!(
+        name = app_instance.name_any(),
+        namespace = app_instance.namespace(),
+        "Cleaning up!"
+    );
+    let jobs: Api<Job> = Api::namespaced(ctx.client.clone(), &app_instance.namespace_any());
+    let apply_job_name = job_name_for(app_instance, "apply");
+    let cleanup_job_name = job_name_for(app_instance, "cleanup");
+
+    match jobs.get_opt(&apply_job_name).await? {
+        Some(apply_job) => {
+            info!("Deleting the running job");
+            delete_job(app_instance, ctx).await?;
+
+            info!("Awaiting termination of {apply_job_name}");
+            let job_uid = apply_job.uid().unwrap();
+            let cond = await_condition(jobs.clone(), &apply_job_name, is_deleted(&job_uid));
+
+            // Cleaning up the job can take some time and is an idempotent action, so we
+            // can requeue if upon failure when an Err is returned.
+            if tokio::time::timeout(Duration::from_secs(120), cond)
+                .await
+                .is_err()
+            {
+                Err(Error::ResourceDeletionTimeout)
+            } else {
+                create_cleanup(app_instance, jobs, &cleanup_job_name, ctx).await
+            }
+        }
+        None => {
+            info!("No Job found for {apply_job_name}, proceeding to cleanup phase");
+            match jobs.get_opt(&cleanup_job_name).await? {
+                Some(_) => {
+                    create_cleanup(app_instance, jobs, &cleanup_job_name, ctx).await?;
+                    delete_cleanup_hack_configmap(app_instance, ctx).await
+                }
+                None => {
+                    delete_cleanup_hack_configmap(app_instance, ctx).await?;
+                    Ok(Action::await_change())
+                }
+            }
+        }
+    }
+}
+
+/// Delete the ConfigMap that was used to prune the applyset.
+///
+/// This diverges slightly from the spawned Job with emit_<command> style
+/// that is used throughout the codebase, as the Job is marked as `Completed`.
+/// This is problematic because we cannot use a `PreStop` hook in order to
+/// run a `kubectl delete configmap` operation.
+///
+/// From the Kubernetes documentation:
+/// A call to the PreStop hook fails if the container is already in a terminated or completed
+/// state
+///
+/// For further details see
+/// <https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks>
+async fn delete_cleanup_hack_configmap(
```

Comment:
This is the small divergence, but if we're okay with this small hack as the final cleanup step then overall I think this PR is in a good state.

By divergence, I mean that it does not use the `emit_*` style for the deletion phase. Locally, this _is_ possible, and is in fact what is done, but as for running this in the controller flow as a separate container or `PreStop` hook I either didn't think it was worthwhile or it didn't work.

1 - As a separate container this would mean an entirely new `Job` (perhaps just a standalone `Pod`, which itself would be hacky anyway) for the `ConfigMap` deletion - which doesn't seem worthwhile compared to the 4 lines here which is relatively succinct and avoids further bloat. Perhaps the standalone `Pod` is okay though, in which case I'll remove this hack.

2 - The "better" option I thought would be using a `PreStop` lifecycle hook; however, from my testing it seems that isn't possible here because we are using a `Job`. Such that the container moves into a `Completed` state on success and `Error`/`Terminated` on failure, this stops the `PreStop` hook from ever running; otherwise we could run the `kubectl delete configmap ...` operation at the end of the cleanup `Job`.

Reply:
I think it's ok for the cleanup of the hack configmap to be done by the controller directly and not with a job